sdk/python/kfp/dsl/_pipeline.py

```diff
@@ -139,3 +140,15 @@ def get_next_group_id(self):
 
     self.group_id += 1
     return self.group_id
+
+  # The following functions are used by the users to customize pipeline level configurations
+  # Note: call these function inside the pipeline python function.
+  @staticmethod
+  def set_imagepullsecret(secret_name):
```

Comment:
Note that from Kubernetes perspective there are not pipelines. 
Image pull secrets are specified per container or pod.

I think we need something like `.apply_to_all_tasks` for pipeline analogous to `.apply` or `task_modifiers`.

Mockup:

    dsl.compiler.Compiler().compile(pipeline_func, task_modifiers=[
        use_gcp_secret('user-sa'),
        set_image_pull_secret('image-pull-secret'),
    ])

Reply:
In fact, the pipelinespec is the pipeline itself. However, maybe it is a good idea to group the user configurable fields into a class. Would that be confusing to the authors?

Reply:
It is not supported in the argo yet, though. 
I thought the first step is to add it globally as the workflow, which is supported currently.

Reply:
I am fine to make pipeline interface as the spec, but it's better to cleanup the interface to only expose user facing APIs and make others as internal as we are going to expose it to the user. WDYT?

Reply:
Make sense, How about rename the non-user facing API function names to internal names starting with "_"?

Reply:
sg

Reply:
> It is not supported in the argo yet, though.

Neither are volumes. But once they add it we might be incompatible.

Reply:
Moving the Pipeline to the compiler might not be ideal because the python SDK(containerOP) depends on the Pipeline class. And, our compiler also depends on the python SDK. This circular dependency does not make much sense to me. However, I'm all for making the Pipeline private.